File: 0146-lru-cache/0146-lru-cache.py

class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None


class LRUCache:
    def __init__(self, capacity):
        self.capacity = capacity
        self.cache = {}
        self.head = None
        self.tail = None

    def get(self, key):
        if key in self.cache:
            node = self.cache[key]
            self._move_to_front(node)
            return node.value
        return -1

    def put(self, key, value):
        if key in self.cache:
            node = self.cache[key]
            node.value = value
            self._move_to_front(node)
        else:
            if len(self.cache) >= self.capacity:
                self._remove_last()
            node = Node(key, value)
            self.cache[key] = node
            self._add_to_front(node)

    def _move_to_front(self, node):
        if node == self.head:
            return
        self._remove_node(node)
        self._add_to_front(node)

    def _remove_node(self, node):
        if node.prev:
            node.prev.next = node.next
        if node.next:
            node.next.prev = node.prev
        if node == self.head:
            self.head = node.next
        if node == self.tail:
            self.tail = node.prev

    def _add_to_front(self, node):
        if self.head:
            self.head.prev = node
        node.next = self.head
        node.prev = None
        self.head = node
        if not self.tail:
            self.tail = node

    def _remove_last(self):
        if self.tail:
            del self.cache[self.tail.key]
            if self.tail.prev:
                self.tail.prev.next = None
            self.tail = self.tail.prev
        if not self.tail:
            self.head = None

File: 0146-lru-cache/0146-lru-cache.cpp
#include <unordered_map>
#include<bits/stdc++.h>
class LRUCache {
private:
    struct Node {
        int key;
        int value;
        Node* prev;
        Node* next;
        
        Node(int k, int v) : key(k), value(v), prev(nullptr), next(nullptr) {}
    };
    
    int capacity;
    std::unordered_map<int, Node*> cache;
    Node* head;
    Node* tail;

    void moveToFront(Node* node) {
        if (node == head) {
            return;
        }
        removeNode(node);
        addToFront(node);
    }

    void removeNode(Node* node) {
        if (node->prev) {
            node->prev->next = node->next;
        } else {
            head = node->next;
        }
        if (node->next) {
            node->next->prev = node->prev;
        } else {
            tail = node->prev;
        }
    }

    void addToFront(Node* node) {
        node->next = head;
        node->prev = nullptr;
        if (head) {
            head->prev = node;
        }
        head = node;
        if (!tail) {
            tail = node;
        }
    }

    void removeLast() {
        if (tail) {
            cache.erase(tail->key);
            Node* prev = tail->prev;
            if (prev) {
                prev->next = nullptr;
            } else {
                head = nullptr;
            }
            tail = prev;
        }
    }

public:
    LRUCache(int capacity) : capacity(capacity), head(nullptr), tail(nullptr) {}
    
    int get(int key) {
        if (cache.find(key) != cache.end()) {
            Node* node = cache[key];
            moveToFront(node);
            return node->value;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            Node* node = cache[key];
            node->value = value;
            moveToFront(node);
        } else {
            if (cache.size() >= capacity) {
                removeLast();
            }
            Node* newNode = new Node(key, value);
            cache[key] = newNode;
            addToFront(newNode);
        }
    }
};
